pkg/tls/tls.go

```diff
@@ -1,91 +1,164 @@
 package tls
 
 import (
-	"bytes"
-	"crypto/ecdsa"
-	"crypto/elliptic"
-	"crypto/rand"
 	"crypto/tls"
 	"crypto/x509"
-	"crypto/x509/pkix"
-	"encoding/pem"
-	"math/big"
-	"net"
-	"time"
+	"errors"
+	"io/ioutil"
+	"os"
 )
 
-// GenerateTLSConfig Setup a bare-bones TLS config for the server
-func GenerateTLSConfig(host ...string) *tls.Config {
-	tlsCert, _ := GenerateCertificate(host...)
-
-	return &tls.Config{
-		Certificates:       []tls.Certificate{tlsCert},
-		ClientSessionCache: tls.NewLRUClientSessionCache(1),
-		NextProtos:         []string{"yomo"},
+// Create server tls config.
+func CreateServerTLSConfig() (*tls.Config, error) {
+	pool, err := getCACertPool()
+	if err != nil {
+		return nil, err
 	}
-}
 
-func GenerateCertificate(host ...string) (tls.Certificate, error) {
-	priv, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	tlsCert, err := getCertAndKey()
 	if err != nil {
-		return tls.Certificate{}, err
+		return nil, err
 	}
 
-	notBefore := time.Now()
-	notAfter := notBefore.Add(time.Hour * 24 * 365)
+	var clientAuth tls.ClientAuthType
+	if isDev() {
+		clientAuth = tls.NoClientCert
+	} else {
+		clientAuth = tls.RequireAndVerifyClientCert
+	}
+
+	return &tls.Config{
+		Certificates: []tls.Certificate{*tlsCert},
+		ClientCAs:    pool,
+		ClientAuth:   clientAuth,
+		NextProtos:   []string{"yomo"},
+	}, nil
+}
 
-	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
-	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+// Create client tls config.
```

Comment:
@wujunzhuo should change comment to `CreateClientTLSConfig create client tls config.`